### DISCUSSION TITLE
Explicitly closing a PDO SQLite connection / file

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#50834] [Q&A] [answered]
I'm splitting a large collection into thousands of small SQLite files, but I'm getting a  "Too Many Files Open" error.

I'm creating the connections via

```php
$cache =  new DoctrineDbalAdapter('sqlite:///' . $fullFilename);
$caches[$lru] = $cache;

// the LRU cache, but the files appear to remain open. Obviously, I'm hacking here 
// trying to figure out how to close the underlying SQLite database.

$lruCache = $caches[$lru];
$lruCache->reset();
$lruCache = null;
unset($lruCache);
$caches[$lru] = null; // make sure the file is closed
unset($caches[$lru]);
```

Is there a way to explicitly close the underlying SQLite file?  A ->destroy() or ->close() method somewhere I can call?

Cross-posted to https://stackoverflow.com/questions/76551082/how-do-close-the-file-associated-with-the-dbal-adapter

**Comment by MatTheCat** [comment]
I guess you can pass your own connection to `DoctrineDbalAdapter` (if you're creating them on-the-fly maybe you can consider the `PdoAdapter`) and close it yourself?

**Comment by tacman** [comment]
Thanks, can you please point me to an example of how to do that?  I really struggling with the relationship between a Connection, PdoAdapter and DoctrineDBalAdapter.

```php
           $pdo = new PdoAdapter('sqlite:///' . $fullFilename); 
           $conn = ...?
           $cache =  new DoctrineDbalAdapter($conn);
 ```

**Reply by MatTheCat** [reply, marked as answer]
```php
$conn = new \PDO("sqlite:{$fullFilename}");
$cache = new PdoAdapter($conn);
// ...
$conn->close();
```

This feels very weird though; why would you want thousands of small SQLite files?

**Comment by tacman** [comment]
```php
$conn = new \PDO("sqlite:{$fullFilename}");
$cache = new PdoAdapter($conn);
// ...
$conn->close(); 
```
Throws a "Call to undefined method PDO::close()".     

I needed the files to be portable (so storing them in PostgreSQL was problematic).  Each "tenant" has their data stored in SQLite files that can be imported and manipulated.

**Reply by MatTheCat** [reply]
Ah yes that would be `unset($conn)`.

Please reply using “write a reply” (not “suggest an answer”) next time 😅
